dlib/control/qlearning_abstract.h

```diff
@@ -0,0 +1,205 @@
+            bool verbose
+        );
+        /*!
+          requires
+            - learning_rate >= 0 and learning_rate <= 1
+            - discount >= 0 and discount <= 1
+            - epsilon >= 0 and epsilon <= 1
+          ensures
+            - #get_learning_rate() == learning_rate
+            - #get_discount() == discount
+            - #get_iterations() == iterations
+            - #get_epsilon() == epsilon
+            - #is_verbose() == verbose
+        !*/
+
+        double get_learning_rate(
+        ) const;
+        /*!
+            ensures
+                - returns the learning rate applied to the learning function.
+        !*/
+
+        void set_learning_rate(
+            double learning_rate
+        );
+        /*!
+            requires
+                - learning_rate >= 0 and learning_rate <= 1.
+            ensures
+                - #get_learning_rate() == learning_rate
+        !*/
+
+        double get_discount(
+        ) const;
+        /*!
+            ensures
+                - returns the discount applied to the learning function.
+        !*/
+
+        void set_discount(
+            double discount
+        );
+        /*!
+            requires
+                - discount >= 0 and discount <= 1.
+            ensures
+                - #get_discount() == discount
+        !*/
+
+        unsigned int get_iterations(
+        ) const;
+        /*!
+            ensures
+                - returns the maximum number of iterations that qlearning will
+                  perform during the training.
+        !*/
+
+        void set_iterations(
+            unsigned int iterations
+        );
+        /*!
+            ensures
+                - #get_iterations() == iterations
+        !*/
+
+        double get_epsilon(
+        ) const;
+        /*!
+            ensures
+                - returns the probability of doing a non-optimal step while training.
+        !*/
+
+        void set_epsilon(
+            double epsilon
+        );
+        /*!
+            requires
+                - epsilon >= 0 and epsilon <= 1.
+            ensures
+                - #get_epsilon() == epsilon
+        !*/
+
+        bool is_verbose(
+        ) const;
+        /*!
+            ensures
+                - returns if the class is verbose or not.
+        !*/
+
+        void be_verbose(
+        );
+        /*!
+            ensures
+                - #is_verbose() == true
+        !*/
+
+        void be_quiet(
+        );
+        /*!
+            ensures
+                - #is_verbose() == false
+        !*/
+
+        template <
+            typename policy_type,
+            typename prng_engine = std::default_random_engine
+            >
+        policy_type train_policy(
+            const policy_type &policy,
+            const prng_engine &gen = prng_engine()
+        ) const;
+        /*!
+            requires
+                - policy is of the form example_policy<model_type>, i.e., an instance of
```

Comment:
This is too general.  You should change this function to take a `policy<model>` which is much more specific.  So far I don't see any reason to not be more specific.  That is, why would a user want to create their own policy object?